src/request/request.rs

```diff
@@ -219,12 +219,13 @@ impl<T> Request<T> {
     }
 
     /// Consumes the request returning a new request with body mapped to the
-    /// return type of the provided closure `map`.
+    /// return type of the provided closure.
     ///
     #[inline]
-    pub fn map<F>(self, map: F) -> Self
+    pub fn map<U, F>(self, map: F) -> Self
     where
-        F: FnOnce(Either<Limited<Incoming>, BoxBody>) -> BoxBody,
+        F: FnOnce(RequestBody) -> U,
+        U: Body<Data = Bytes, Error = BoxError> + Send + Sync + 'static,
```

Comment:
Calling map no longer requires the caller to manually box bodies. It does result in a layer of indirection for each time that a body is boxed.

This is a better approach IMO as the implementation details of request and response bodies do not leak into userland. It also feels more correct since the body moves into the mapped type with it's container.

Reply:
This does allow users to create nasty chains of self-referential body types but in practice it is highly unlikely that mapping a request or response more than once is necessary.